linera-chain/src/manager.rs

```diff
@@ -424,51 +426,60 @@ where
         key_pair: Option<&KeyPair>,
         local_time: Timestamp,
         blobs: BTreeMap<BlobId, Blob>,
-    ) -> Result<Option<ValidatedOrConfirmedVote>, ViewError> {
+    ) -> Result<Option<ValidatedOrConfirmedVote>, ChainError> {
         let round = proposal.content.round;
+        if key_pair.is_some() && round < self.current_round() {
+            return Ok(None);
+        }
 
         // If the validated block certificate is more recent, update our locked block.
         if let Some(lite_cert) = &proposal.validated_block_certificate {
             if self
                 .locked
                 .get()
                 .as_ref()
-                .map_or(true, |locked| locked.round < lite_cert.round)
+                .map_or(true, |locked| locked.round() < lite_cert.round)
             {
                 let value = Hashed::new(ValidatedBlock::new(executed_block.clone()));
                 if let Some(certificate) = lite_cert.clone().with_value(value) {
-                    self.set_locked(certificate, blobs)?;
+                    self.set_locked(LockedBlock::Regular(certificate), blobs)?;
                 }
             }
+        } else if round.is_fast() {
+            // The fast block also counts as locked.
+            self.set_locked(LockedBlock::Fast(proposal.clone()), blobs)?;
         }
 
+        let Some(key_pair) = key_pair else {
+            // Record the proposed block, in case it affects the current round number.
+            self.set_proposed(proposal);
+            self.update_current_round(local_time);
+            return Ok(None);
+        };
+
+        self.check_proposal_round(&proposal)?;
```

Comment:
So we check the proposal's vailidity w.r.t. to the round only when we're actively validating? i.e. shouldn't this check be before the `let Some(key_pair) = key_pair else { ... }` which will return early?

Reply:
No, because even if the proposal isn't from the current round, on the client side we still want to add it to the `proposed` field so that in multi leader rounds we know we shouldn't make a proposal in the same round ourselves. (This _partially_ addresses https://github.com/linera-io/linera-protocol/issues/2971.)

Reply:
That's confusing TBH. That the logic here is makes these distinction b/c of some hidden requirements (i.e. being invoked on client vs server side).

Reply:
I agree. As I mentioned on Slack, I think we should make that distinction more explicit. But I think that will be a bigger change, and I'm not sure if it wouldn't make this PR even harder to review. Happy to give it a try, though, if you prefer.

Reply:
No, that's fine. Just maybe leave that part in addition to the comment above (in the else { ... } clause).

Reply:
Actually, I'll also add another check so that we only add the proposal if it's before the single leader rounds. For those we don't need it, and we also shouldn't accept it before the round has begun. Sorry for the confusion! I hope we can separate some of the client vs. validator logic soon, and clarify all of this.